support dhcp{4,6}-overrides in networkd renderer

adobley opened the pull request:
## Proposed Commit Message
<!-- Include a proposed commit message because all PRs are squash merged -->

```
support dhcp{4,6}-overrides in networkd renderer
```

## Additional Context
<!-- If relevant -->

## Test Steps
<!-- Please include any steps necessary to verify (and reproduce if
this is a bug fix) this change on a live deployed system,
including any necessary configuration files, user-data,
setup, and teardown. Scripts used may be attached directly to this PR. -->

Create the metadata file for the VM `metadata.yaml`:
```
instance-id: cloud-vm
local-hostname: cloud-vm
network:
  version: 2
  ethernets:
    nics:
      match:
        name: ens*
      dhcp4: true
      dhcp4-overrides:
        use-dns: false
      nameservers:
        addresses: [8.8.8.8, 8.8.4.4]
```

Create the userdata file `userdata.yaml`:
```
#cloud-config

users:
- default
- name: ${YOUR_USER}
    primary_group: ${YOUR GROUP}
    sudo: ALL=(ALL) NOPASSWD:ALL
    groups: sudo, wheel
    lock_passwd: true
    ssh_authorized_keys:
    - ${YOUR_SSH_PUB_KEY}
```

Create a powered-off Photon VM in vSphere:
```
govc import.ova photon-3-kube.ova
govc vm.markastemplate photon-3-kube
govc vm.clone -on=false -vm /dc0/vm/photon-3-kube cloud-vm
```

Set the guestinfo on the VM while it is powered off:
```
export GUESTINFO_USERDATA="$(cat userdata.yaml | base64 -w0)"
export GUESTINFO_METADATA="$(cat metadata.yaml | base64 -w0)"

govc vm.change -vm /dc0/vm/cloud-vm \
	-e guestinfo.userdata="${GUESTINFO_USERDATA}" \
	-e guestinfo.userdata.encoding="base64" \
	-e guestinfo.metadata="${GUESTINFO_METADATA}" \
	-e guestinfo.metadata.encoding="base64"
```

Power on the VM:
```
govc vm.power -on /dc0/vm/humalog
```

SSH to the VM and check the resulting network config is as expected:
```
ssh ${YOUR_USER}@${CLOUD_VM_IP}
networkctl status
```

See that only the configured DNS servers are in the list and no DNS servers came from DHCP.
This is just one of the several flags that could be toggled but gives an idea if it's wired up correctly.


## Checklist:
<!-- Go over all the following points, and put an `x` in all the boxes
that apply. -->
 - [x] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/topics/contributing.html)
 - [x] I have updated or added any unit tests accordingly
 - [x] I have updated or added any documentation accordingly

aciba90 commented on cloudinit/net/networkd.py:
```diff
@@ -271,7 +298,23 @@ def _render_content(self, ns):
                 if name in ns.config["ethernets"]:
-                    self.dhcp_domain(ns.config["ethernets"][name], cfg)
+                    device = ns.config["ethernets"][name]
+                    for version in ["4", "6"]:
```
This check is specific for VMware config. The following diff adds a comment underlying that and simplifying the if-condition. Feel free to rephrase it.

```diff
diff --git a/cloudinit/net/networkd.py b/cloudinit/net/networkd.py
index d9ad1fc8c..c54352956 100644
--- a/cloudinit/net/networkd.py
+++ b/cloudinit/net/networkd.py
@@ -299,12 +299,13 @@ class Renderer(renderer.Renderer):
                             break
                 if name in ns.config["ethernets"]:
                     device = ns.config["ethernets"][name]
+                    # dhcp{version}domain are extra keys only present in
+                    # VMware config.
                     for version in ["4", "6"]:
                         if (
                             f"dhcp{version}domain" in device
-                            and f"dhcp{version}-overrides" in device
                             and "use-domains"
-                            in device[f"dhcp{version}-overrides"]
+                            in device.get(f"dhcp{version}-overrides", {})
                         ):
                             exception = (
                                 f"{name} has both dhcp{version}domain"

```

adobley replied:
Thanks for the tip! I haven't written python in a while. Glad there was a better way :)

aciba90 commented on tests/unittests/net/test_networkd.py:
```diff
@@ -71,5 +183,89 @@ def test_networkd_render_with_set_name(self):
+
+    @pytest.mark.parametrize("dhcp_version", [("4"), ("6")])
+    def test_networkd_render_dhcp_domains_vs_overrides(self, dhcp_version):
+        expectedException = (
```
In general, we use snake-case to name vars. Could we rename `expectedException` to `expected_exception`?

adobley replied:
Whoops! Thanks for catching that mistake.

adobley commented on tests/unittests/net/test_networkd.py:
```diff
@@ -53,6 +56,115 @@
+      match:
+        macaddress: '00:11:22:33:44:55'
+      nameservers:
+        addresses: [8.8.8.8,2001:4860:4860::8888]
```
Fixed, thanks!

adobley commented on cloudinit/net/networkd.py:
```diff
@@ -216,6 +218,31 @@ def parse_dns(self, iface, cfg, ns):
+            "UseRoutes": "use-routes",
+        }
+
+        if "dhcp4-overrides" in device and dhcp in ["yes", "ipv4"]:
```
We took a slightly different approach of making this function parameterized with that data and using the existing for loop we had to do the validation on the conflicting keys. Let me know if there are any concerns :)

adobley commented:
Hi @aciba90! Thanks for all your feedback! We fixed up the suggestions and added some docs on the fields. I'd love to hear any thoughts you have.

Would it be better to just link out to the netplan docs for the explanations of some of the properties?
Is there a better way we can show that note/warning in the docs for the renderer limitations?

We weren't too sure where a good place to add a log that alerts users when `dhcp{4,6}-overrides` are set and they are not using the `networkd` renderer. I didn't see any obvious place doing a high level validation like that, but maybe I missed it. We could add something just before the render gets called, if we have enough information then.

aciba90 commented on doc/rtd/topics/network-config-format-v2.rst:
```diff
@@ -177,6 +177,72 @@ Enable DHCP for IPv4. Off by default.
 
 Enable DHCP for IPv6. Off by default.
 
+**dhcp4-overrides**: *<(mapping)>*
```
The following diff simplifies and improves the documentation. By:

* Not documenting the sub-properties, but referring to netplan's doc.
* Grouping dhcp4-overrides and dhcp6-overrides in one section.
* Clarify that with v2 netconfs and netplan as renderer, the overrides are also applied.

Please, feel free to modify/rephrase it.

```diff
diff --git a/doc/rtd/topics/network-config-format-v2.rst b/doc/rtd/topics/network-config-format-v2.rst
index 8597cc32b..b00d4110e 100644
--- a/doc/rtd/topics/network-config-format-v2.rst
+++ b/doc/rtd/topics/network-config-format-v2.rst
@@ -8,6 +8,8 @@ version 2 format defined for the `netplan`_ tool.  Cloud-init supports
 both reading and writing of Version 2; the latter support requires a
 distro with `netplan`_ present.
 
+.. _Netplan Passthrough:
+
 Netplan Passthrough
 -------------------
 
@@ -177,62 +179,32 @@ Enable DHCP for IPv4. Off by default.
 
 Enable DHCP for IPv6. Off by default.
 
-**dhcp4-overrides**: *<(mapping)>*
-
-Some DHCP behaviors can be overridden. Overrides will only have an effect if
-the corresponding DHCP type is enabled. Optional.
-
-``hostname`` *<(scalar)>* sets the hostname which is sent to the DHCP server,
-instead of the machine's hostname. Optional.
-``route-metric`` *<(scalar)>* sets the default metric for routes that are
-automatically added. This allows prioritizing routes for devices. Optional.
-``send-hostname`` *<(bool)>* when ``true``, the machine's hostname will be sent
-to the DHCP server. Default: ``true``. Optional.
-``use-dns`` *<(bool)>* when ``true``, the DNS servers received from DHCP will
-be used and take precedence over configured ones. Default: ``true``. Optional.
-``use-hostname`` *<(bool)>* when ``true``,  the hostname received from DHCP
-will be set as the transient hostname of the machine. Default: ``true``. Optional.
-``use-mtu`` *<(bool)>* when ``true``, the MTU received from DHCP will be used
-as the MTU for the network interface. Default: ``true``. Optional.
-``use-ntp`` *<(bool)>* when ``true``, the NTP servers received from DHCP will
-be used by systemd-timesyncd and take precendence over configured ones.
-Default: ``true``. Optional.
-``use-domains`` *<(scalar)>* takes a boolean or the value ``route``. When
-``true``, the domain name received from DHCP will be used as the DNS search
-domain for this network interface. When ``route``, the domain name received
-from DHCP will be used for routing DNS only, not for searching. Optional.
-
-NOTE: This is currently only supported for the ``networkd`` renderer. As of
-writing this the only known usage of this renderer is with the VMware data
-source.
+**dhcp4-overrides** and **dhcp6-overrides**: *<(mapping)>*
 
-Example: ::
+DHCP behavior overrides. Overrides will only have an effect if
+the corresponding DHCP type is enabled. Refer to `netplan#dhcp-overrides`_
+for more documentation.
 
-  dhcp4-overrides:
-    hostname: hal
-    route-metric: 1100
-    send-hostname: false
-    use-dns: false
-    use-domains: false
-    use-hostname: false
-    use-mtu: false
-    use-ntp: false
-    use-routes: false
-
-**dhcp6-overrides**: *<(mapping)>*
+.. note::
 
-Some DHCP behaviors can be overridden. Overrides will only have an effect if
-the corresponding DHCP type is enabled. Optional.
+  These properties are only consumed on ``netplan`` and ``networkd``
+  renderers.
 
-See ``dhcp4-overrides`` above for an explanation of the fields.
+The ``netplan`` renderer :ref:`passes through <Netplan Passthrough>`
+everything and the ``networkd`` renderer consumes the following sub-properties:
 
-NOTE: This is currently only supported for the ``networkd`` renderer. As of
-writing this the only known usage of this renderer is with the VMware data
-source.
+* ``hostname``
+* ``route-metric``
+* ``send-hostname``
+* ``use-dns``
+* ``use-domains``
+* ``use-hostname``
+* ``use-mtu``
+* ``use-ntp``
 
 Example: ::
 
-  dhcp6-overrides:
+  dhcp4-overrides:
     hostname: hal
     route-metric: 1100
     send-hostname: false
@@ -593,4 +565,5 @@ This is a complex example which shows most available features: ::
         dhcp4: yes
 
 .. _netplan: https://netplan.io
+.. _netplan#dhcp-overrides: https://netplan.io/reference#dhcp-overrides
 .. vi: textwidth=79

```

adobley replied:
Thanks! This patch looks great!

aciba90 reviewed (changes requested):
Thanks, @adobley, for the improvements and documentation.

> Would it be better to just link out to the netplan docs for the explanations of some of the properties?

I left a comment suggesting changes in the documentation. I think is better to refer to netplan's doc, rather than duplicating sub-props documentation.

> Is there a better way we can show that note/warning in the docs for the renderer limitations?

Furthermore, I included a `note` sphinx directive to better signalize that fact.

> We weren't too sure where a good place to add a log that alerts users when dhcp{4,6}-overrides are set and they are not using the networkd renderer. I didn't see any obvious place doing a high level validation like that, but maybe I missed it. We could add something just before the render gets called, if we have enough information then.

Maybe it's enough with the outlined docs, and we can leave the warning as a future improvement.

adobley commented:
Thanks for all your help @aciba90! I think we've got everything. :)

aciba90 reviewed (approved):
LGTM! Thanks, @adobley, for this addition!

TheRealFalcon reviewed (commented):
If I use this config:
```
network:
  version: 2
  ethernets:
    eth0:
      dhcp4: true
      dhcp4-overrides:
        hostname: hal
        route-metric: 1100
        send-hostname: false
        use-dns: false
        use-domains: false
        use-hostname: false
        use-mtu: false
        use-ntp: false
        use-routes: false
      dhcp6: true
      dhcp6-overrides:
        hostname: hal
        route-metric: 1100
        send-hostname: false
        use-dns: false
        use-domains: false
        use-hostname: false
        use-mtu: false
        use-ntp: false
        use-routes: false
      match:
        macaddress: "00:11:22:33:44:55"
      nameservers:
        addresses: ["8.8.8.8", "2001:4860:4860::8888"]
```

I get a network file with this:
```
[DHCPv4]
Hostname=hal
RouteMetric=1100
SendHostname=False
UseDNS=False
UseDomains=False
UseHostname=False
UseMTU=False
UseNTP=False
UseRoutes=False

[DHCPv6]
Hostname=hal
RouteMetric=1100
SendHostname=False
UseDNS=False
UseDomains=False
UseHostname=False
UseMTU=False
UseNTP=False
UseRoutes=False

[Match]
MACAddress=00:11:22:33:44:55
Name=eth0

[Network]
DHCP=yes
DNS=8.8.8.8 2001:4860:4860::8888
```

But looking at [the networkd documentation](https://www.freedesktop.org/software/systemd/man/systemd.network.html#%5BDHCPv6%5D%20Section%20Options), it appears that most of those options aren't valid for `[DHCPv6]` using networkd.

Is the resulting `[DHCPv6]` section actually valid?

adobley commented on doc/rtd/topics/network-config-format-v2.rst:
```diff
@@ -177,6 +179,42 @@ Enable DHCP for IPv4. Off by default.
+* ``use-hostname``
+* ``use-mtu``
+* ``use-ntp``
+
```
Thanks! We made some broader changes to the docs to cover the difference in configuration options between netplan and networkd. We were sure to add this when we made that change. :)

adobley commented:
@TheRealFalcon Thanks for catching that mistake! It does look like those properties are ignored for DHCPv6. We updated the code to reflect this and took a first pass at some tweaks to the documentation. Any feedback you have is appreciated!

TheRealFalcon commented on cloudinit/net/networkd.py:
```diff
@@ -271,7 +299,26 @@ def _render_content(self, ns):
+                                f" and dhcp{version}-overrides.use-domains"
+                                f" configured. Use one"
+                            )
+                            LOG.critical(exception)
```
Nit: An exception raised from here will make it into the logs, so we don't need an additional log here too.

adobley replied:
Fixed. Thanks!

TheRealFalcon reviewed (approved):
LGTM! One nit inline and we need to fix the black failure. Once all the checks pass, I can merge.

TheRealFalcon commented:
Closing and re-opening to trigger Travis

TheRealFalcon reviewed (approved):
This looks good now. Thanks for the contribution!
